### PR TITLE
Fix additional properties can not be added for Cliams in External Claim dialects

### PR DESCRIPTION
**Pull request by sahanruwanga**
### Proposed changes in this pull request

- Resolves [wso2/product-is/issue-3821](https://github.com/wso2/product-is/issues/3821)
- **There is a Migration Impact**
  A column name of an existing table is changed with this PR
  Table Name: **IDN_CLAIM_PROPERTY**
  Old (existing) Column Name: **LOCAL_CLAIM_ID**
  New Column Name: **CLAIM_ID**
Therefore, this change should be included in **Migration Scripts**

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]
Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

#### General

  - [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
  - [ ] **Is the PR labeled correctly?**

#### Functionality

   - [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
   - [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

   - [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
   - [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
   - [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be marked for translation.

#### Tests

   - [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
   - [ ] **If this is a bug fix, are tests for the issue in place?** There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
   - [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

   - [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
   - [ ] **Are all UI and API inputs run through forms or serializers?**
   - [ ] **Are all external inputs validated and sanitized appropriately?**
   - [ ] **Does all branching logic have a default case?**
   - [ ] **Does this solution handle outliers and edge cases gracefully?**
   - [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

   - [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
   - [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
   - [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt.ui/pom.xml [hunk 1]
```diff
@@ -56,6 +56,11 @@
             <groupId>org.wso2.orbit.org.owasp.encoder</groupId>
             <artifactId>encoder</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.identity.framework</groupId>
+            <artifactId>org.wso2.carbon.identity.claim.metadata.mgt.stub</artifactId>
+            <version>5.12.154-SNAPSHOT</version>
```
Comment (madurangasiriwardena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt.ui/pom.xml]: remove snapshot dependency

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/DefaultClaimMetadataStore.java [hunk 2]
```diff
@@ -208,6 +183,36 @@ private void init(ClaimConfig claimConfig, int tenantId) {
 
     }
 
+    private Map<String, String> getClaimProperties(ClaimConfig claimConfig, ClaimKey claimKey){
```
Comment (madurangasiriwardena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/DefaultClaimMetadataStore.java]: Shall we change the method name to a more meaningful name?

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ClaimDAO.java [hunk 3]
```diff
@@ -173,4 +173,80 @@ public int getClaimId(Connection connection, String claimDialectURI, String clai
...
+            prepStmt.setInt(1, claimId);
+            prepStmt.setInt(2, tenantId);
+            rs = prepStmt.executeQuery();
+
+            while (rs.next()) {
+                String claimPropertyName = rs.getString(SQLConstants.PROPERTY_NAME_COLUMN);
+                String claimPropertyValue = rs.getString(SQLConstants.PROPERTY_VALUE_COLUMN);
+
+                claimProperties.put(claimPropertyName, claimPropertyValue);
+            }
```
Comment (madurangasiriwardena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ClaimDAO.java]: Need to commit the connection.
To close connection, prepared statement and result set better use try with resource.
Do this for all the newly added DAO methods.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java [hunk 4]
```diff
@@ -284,7 +295,15 @@ public void testConvertExternalClaimToExternalClaimDTO() throws Exception {
         Assert.assertEquals(externalClaimDTO.getExternalClaimDialectURI(), externalClaim.getClaimDialectURI());
         Assert.assertEquals(externalClaimDTO.getExternalClaimURI(), externalClaim.getClaimURI());
         Assert.assertEquals(externalClaimDTO.getMappedLocalClaimURI(), externalClaim.getMappedLocalClaim());
+        Assert.assertEquals(externalClaimDTO.getClaimProperties().length, externalClaim.getClaimProperties().size());
+
+        for (int i = 0; i < externalClaimDTO.getClaimProperties().length; i++) {
```
Comment (madurangasiriwardena) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtilsTest.java]: Use for each loop.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wso2.carbon.identity.claim.metadata.mgt.dao;
+
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import org.wso2.carbon.identity.claim.metadata.mgt.exception.ClaimMetadataException;
+import org.wso2.carbon.identity.claim.metadata.mgt.model.*;
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java]: It's better to remove wildcard imports.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+import org.wso2.carbon.identity.claim.metadata.mgt.util.ClaimConstants;
+import org.wso2.carbon.identity.common.testng.WithCarbonHome;
+import org.wso2.carbon.identity.common.testng.WithH2Database;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.testng.Assert.*;
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java]: It's better to remove wildcard imports.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+
+@Test
+@WithH2Database(jndiName = "jdbc/WSO2CarbonDB", files = {"dbScripts/claim_properties.sql"})
+@WithCarbonHome
+public class ExternalClaimDAOTest {
+
+    private static final int TENANT_ID = -1234;
+
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java]: Remove unnecessary newlines.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+    private static final String EXTERNAL_CLAIM_DIALECT_URI = "TestExternalClaimDialect";
+
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java]: Remove unnecessary newlines.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+    AttributeMapping attributeMapping1;
+    AttributeMapping attributeMapping2;
+    AttributeMapping attributeMapping3;
+
+    Map<String, String> claimProperties1;
+    Map<String, String> claimProperties2;
+    Map<String, String> claimProperties3;
+
+
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java]: Remove unnecessary newlines.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+        claimDialectDAO2.addClaimDialect(claimDialect2, TENANT_ID);
+
+        LocalClaimDAO localClaimDAO = new LocalClaimDAO();
+        localClaimDAO.addLocalClaim((LocalClaim) localClaim, tenantId);
+
+        ExternalClaimDAO externalClaimDAO = new ExternalClaimDAO();
+
+        externalClaimDAO.addExternalClaim((ExternalClaim) externalClaim, tenantId);
+
+        assertNotNull(externalClaimDAO.getExternalClaims(((ExternalClaim) externalClaim).getClaimDialectURI(), tenantId),
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java]: Check whether this exceeds 120 character limit.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAOTest.java [hunk 11]
```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAOTest.java]: Fix licence year.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAOTest.java [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+import org.wso2.carbon.identity.claim.metadata.mgt.util.ClaimConstants;
+import org.wso2.carbon.identity.common.testng.WithCarbonHome;
+import org.wso2.carbon.identity.common.testng.WithH2Database;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.testng.Assert.*;
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAOTest.java]: Remove wild card imports.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAOTest.java [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wso2.carbon.identity.claim.metadata.mgt.dao;
+
+import org.testng.annotations.*;
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAOTest.java]: Remove wildcard imports.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAOTest.java [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+                        TEST_LOCAL_TENANT_ID
+                },
+                {
+                        localClaim3,
+                        TEST_LOCAL_TENANT_ID
+                },
+
+        };
+    }
+}
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/LocalClaimDAOTest.java]: Add a newline.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/resources/dbScripts/claim_properties.sql [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+CREATE TABLE IF NOT EXISTS IDN_CLAIM_MAPPING (
+  ID INTEGER NOT NULL AUTO_INCREMENT,
+  EXT_CLAIM_ID INTEGER NOT NULL,
+  MAPPED_LOCAL_CLAIM_ID INTEGER NOT NULL,
+  TENANT_ID INTEGER NOT NULL,
+  PRIMARY KEY (ID),
+  FOREIGN KEY (EXT_CLAIM_ID) REFERENCES IDN_CLAIM(ID) ON DELETE CASCADE,
+  FOREIGN KEY (MAPPED_LOCAL_CLAIM_ID) REFERENCES IDN_CLAIM(ID) ON DELETE CASCADE,
+  CONSTRAINT EXT_TO_LOC_MAPPING_CONSTRN UNIQUE (EXT_CLAIM_ID, TENANT_ID),
+);
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/resources/dbScripts/claim_properties.sql]: Add a newline.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/resources/repository/conf/carbon.xml [hunk 16]
```diff
@@ -0,0 +1,684 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+<!--
+ Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/resources/repository/conf/carbon.xml]: Fix license year.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/resources/repository/conf/identity/identity.xml [hunk 17]
```diff
@@ -0,0 +1,976 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+<!--
+~ Copyright (c) 2011, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/resources/repository/conf/identity/identity.xml]: Fix license year.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/model/ExternalClaim.java [hunk 18]
```diff
@@ -27,7 +29,18 @@ public ExternalClaim(String claimDialectURI, String claimURI, String mappedLocal
         this.mappedLocalClaim = mappedLocalClaimURI;
     }
 
+    public ExternalClaim(String claimDialectURI, String claimURI, String mappedLocalClaimURI,
+                         Map<String, String> claimProperties) {
+        super(claimDialectURI, claimURI, claimProperties);
+
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/model/ExternalClaim.java]: Remove unnecessary newlines.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtils.java [hunk 19]
```diff
@@ -160,6 +160,22 @@ public static ExternalClaimDTO convertExternalClaimToExternalClaimDTO(ExternalCl
...
+
+        int j = 0;
+        for (Map.Entry<String, String> claimPropertyEntry : claimProperties.entrySet()) {
+            ClaimPropertyDTO claimProperty = new ClaimPropertyDTO();
+            claimProperty.setPropertyName(claimPropertyEntry.getKey());
+            claimProperty.setPropertyValue(claimPropertyEntry.getValue());
+            claimPropertyDTOs[j] = claimProperty;
+            j++;
+        }
+
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/main/java/org/wso2/carbon/identity/claim/metadata/mgt/util/ClaimMetadataUtils.java]: Remove unnecessary newlines.

File: components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java [hunk 20]
```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (ashensw) [inline components/claim-mgt/org.wso2.carbon.identity.claim.metadata.mgt/src/test/java/org/wso2/carbon/identity/claim/metadata/mgt/dao/ExternalClaimDAOTest.java]: Fix license year.

**Comment by madurangasiriwardena** [comment]
Add migration changes as well.

**Pull request merged**